deepmd/cluster/local.py

```diff
@@ -1,10 +1,58 @@
-"""Get local GPU resources from `CUDA_VISIBLE_DEVICES` enviroment variable."""
-
+"""Get local GPU resources."""
+import logging
 import os
 import socket
+
+import GPUtil
+
+from deepmd.env import tf
 from typing import List, Tuple, Optional
 
-__all__ = ["get_resource"]
+__all__ = ["get_gpus", "get_resource"]
+
+log = logging.getLogger(__name__)
+
+
+def get_gpus():
+    """Get available IDs of GPU cards at local.
+    These IDs are valid when used as the TensorFlow device ID.
+
+    Returns:
+    -------
+    Optional[List[int]]
+        List of available GPU IDs. Otherwise, None.
+    """
+    # TODO: Create a pull request of `GPUtil` to cover ROCM devices.
+    # Currently, even if None is returned, a ROCM device is still visible in TensorFlow.
+    available = GPUtil.getGPUs()
```

Comment:
Can we use [`tf.python.client.device_lib.list_local_devices()`](https://github.com/tensorflow/tensorflow/blob/e0111ef751ed65406117680efed4e48e0b447503/tensorflow/python/client/device_lib.py#L27-L46) instead? Here is how to use it:
```py
>>> from tensorflow.python.client.device_lib import list_local_devices
>>> gpus = [x for x in list_local_devices() if x.device_type == 'GPU']
>>> len(gpus)
2
>>> [gpu.physical_device_desc for gpu in gpus]
['device: 0, name: NVIDIA GeForce RTX 2080 SUPER, pci bus id: 0000:01:00.0, compute capability: 7.5', 'device: 1, name: NVIDIA GeForce RTX 2080 SUPER, pci bus id: 0000:02:00.0, compute capability: 7.5']
```
So we don't need to install a more third-party package any more.

Reply:
No, we cannot use this TensorFlow API because it will automatically register all visible GPUs to TensorFlow device manager. As a result, we cannot limit to certain device using `tf.ConfigProto.gpu_options.visible_device_list` when creating a TensorFlow session any more.
It can be reproduce by the following code.
```
>>> from tensorflow.python.client.device_lib import list_local_devices
2021-07-31 11:13:53.339399: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
>>> gpus = [x for x in list_local_devices() if x.device_type == 'GPU']
2021-07-31 11:13:58.062063: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-07-31 11:13:58.063381: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcuda.so.1
2021-07-31 11:14:00.630159: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 0 with properties:
pciBusID: 0000:0e:00.0 name: A100-SXM4-40GB computeCapability: 8.0
coreClock: 1.41GHz coreCount: 108 deviceMemorySize: 39.59GiB deviceMemoryBandwidth: 1.41TiB/s
2021-07-31 11:14:00.633785: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 1 with properties:
pciBusID: 0000:13:00.0 name: A100-SXM4-40GB computeCapability: 8.0
coreClock: 1.41GHz coreCount: 108 deviceMemorySize: 39.59GiB deviceMemoryBandwidth: 1.41TiB/s
2021-07-31 11:14:00.637093: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 2 with properties:
pciBusID: 0000:4b:00.0 name: A100-SXM4-40GB computeCapability: 8.0
coreClock: 1.41GHz coreCount: 108 deviceMemorySize: 39.59GiB deviceMemoryBandwidth: 1.41TiB/s
2021-07-31 11:14:00.640675: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 3 with properties:
pciBusID: 0000:51:00.0 name: A100-SXM4-40GB computeCapability: 8.0
coreClock: 1.41GHz coreCount: 108 deviceMemorySize: 39.59GiB deviceMemoryBandwidth: 1.41TiB/s
2021-07-31 11:14:00.644086: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 4 with properties:
pciBusID: 0000:93:00.0 name: A100-SXM4-40GB computeCapability: 8.0
coreClock: 1.41GHz coreCount: 108 deviceMemorySize: 39.59GiB deviceMemoryBandwidth: 1.41TiB/s
2021-07-31 11:14:00.647476: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 5 with properties:
pciBusID: 0000:99:00.0 name: A100-SXM4-40GB computeCapability: 8.0
coreClock: 1.41GHz coreCount: 108 deviceMemorySize: 39.59GiB deviceMemoryBandwidth: 1.41TiB/s
2021-07-31 11:14:00.650799: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 6 with properties:
pciBusID: 0000:cb:00.0 name: A100-SXM4-40GB computeCapability: 8.0
coreClock: 1.41GHz coreCount: 108 deviceMemorySize: 39.59GiB deviceMemoryBandwidth: 1.41TiB/s
2021-07-31 11:14:00.654239: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 7 with properties:
pciBusID: 0000:d0:00.0 name: A100-SXM4-40GB computeCapability: 8.0
coreClock: 1.41GHz coreCount: 108 deviceMemorySize: 39.59GiB deviceMemoryBandwidth: 1.41TiB/s
2021-07-31 11:14:00.654273: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
2021-07-31 11:14:00.656455: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublas.so.11
2021-07-31 11:14:00.656491: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublasLt.so.11
2021-07-31 11:14:00.657439: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcufft.so.10
2021-07-31 11:14:00.657668: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcurand.so.10
2021-07-31 11:14:00.659838: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusolver.so.10
2021-07-31 11:14:00.660404: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusparse.so.11
2021-07-31 11:14:00.660540: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudnn.so.8
2021-07-31 11:14:00.711591: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1862] Adding visible gpu devices: 0, 1, 2, 3, 4, 5, 6, 7
2021-07-31 11:14:00.711641: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
2021-07-31 11:14:04.131423: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1261] Device interconnect StreamExecutor with strength 1 edge matrix:
2021-07-31 11:14:04.131521: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1267]      0 1 2 3 4 5 6 7
2021-07-31 11:14:04.131532: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1280] 0:   N Y Y Y Y Y Y Y
2021-07-31 11:14:04.131539: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1280] 1:   Y N Y Y Y Y Y Y
2021-07-31 11:14:04.131545: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1280] 2:   Y Y N Y Y Y Y Y
2021-07-31 11:14:04.131553: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1280] 3:   Y Y Y N Y Y Y Y
2021-07-31 11:14:04.131560: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1280] 4:   Y Y Y Y N Y Y Y
2021-07-31 11:14:04.131567: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1280] 5:   Y Y Y Y Y N Y Y
2021-07-31 11:14:04.131574: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1280] 6:   Y Y Y Y Y Y N Y
2021-07-31 11:14:04.131581: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1280] 7:   Y Y Y Y Y Y Y N
2021-07-31 11:14:04.157448: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1406] Created TensorFlow device (/device:GPU:0 with 37571 MB memory) -> physical GPU (device: 0, name: A100-SXM4-40GB, pci bus id: 0000:0e:00.0, compute capability: 8.0)
2021-07-31 11:14:04.165562: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1406] Created TensorFlow device (/device:GPU:1 with 37571 MB memory) -> physical GPU (device: 1, name: A100-SXM4-40GB, pci bus id: 0000:13:00.0, compute capability: 8.0)
2021-07-31 11:14:04.171861: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1406] Created TensorFlow device (/device:GPU:2 with 37571 MB memory) -> physical GPU (device: 2, name: A100-SXM4-40GB, pci bus id: 0000:4b:00.0, compute capability: 8.0)
2021-07-31 11:14:04.189782: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1406] Created TensorFlow device (/device:GPU:3 with 37571 MB memory) -> physical GPU (device: 3, name: A100-SXM4-40GB, pci bus id: 0000:51:00.0, compute capability: 8.0)
2021-07-31 11:14:04.209425: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1406] Created TensorFlow device (/device:GPU:4 with 37571 MB memory) -> physical GPU (device: 4, name: A100-SXM4-40GB, pci bus id: 0000:93:00.0, compute capability: 8.0)
2021-07-31 11:14:04.220758: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1406] Created TensorFlow device (/device:GPU:5 with 37571 MB memory) -> physical GPU (device: 5, name: A100-SXM4-40GB, pci bus id: 0000:99:00.0, compute capability: 8.0)
2021-07-31 11:14:04.227241: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1406] Created TensorFlow device (/device:GPU:6 with 37571 MB memory) -> physical GPU (device: 6, name: A100-SXM4-40GB, pci bus id: 0000:cb:00.0, compute capability: 8.0)
2021-07-31 11:14:04.237115: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1406] Created TensorFlow device (/device:GPU:7 with 37571 MB memory) -> physical GPU (device: 7, name: A100-SXM4-40GB, pci bus id: 0000:d0:00.0, compute capability: 8.0)
>>> import tensorflow.compat.v1 as tf
>>> config = tf.ConfigProto()
>>> config.gpu_options.visible_device_list = u'1'
>>> sess = tf.Session(config=config)
2021-07-31 11:14:39.150037: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-07-31 11:14:39.152410: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 0 with properties:
pciBusID: 0000:13:00.0 name: A100-SXM4-40GB computeCapability: 8.0
coreClock: 1.41GHz coreCount: 108 deviceMemorySize: 39.59GiB deviceMemoryBandwidth: 1.41TiB/s
2021-07-31 11:14:39.152445: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
2021-07-31 11:14:39.152472: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublas.so.11
2021-07-31 11:14:39.152485: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublasLt.so.11
2021-07-31 11:14:39.152495: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcufft.so.10
2021-07-31 11:14:39.152506: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcurand.so.10
2021-07-31 11:14:39.152517: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusolver.so.10
2021-07-31 11:14:39.152529: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusparse.so.11
2021-07-31 11:14:39.152541: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudnn.so.8
2021-07-31 11:14:39.156594: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1862] Adding visible gpu devices: 1
2021-07-31 11:14:39.156639: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1261] Device interconnect StreamExecutor with strength 1 edge matrix:
2021-07-31 11:14:39.156648: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1267]      1
2021-07-31 11:14:39.156661: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1280] 1:   N
2021-07-31 11:14:39.156737: E tensorflow/core/common_runtime/session.cc:91] Failed to create session: Already exists: TensorFlow device (GPU:0) is being mapped to multiple CUDA devices (1 now, and 0 previously), which is not supported. This may be the result of providing different GPU configurations (ConfigProto.gpu_options, for example different visible_device_list) when creating multiple Sessions in the same process. This is not  currently supported, see https://github.com/tensorflow/tensorflow/issues/19083
2021-07-31 11:14:39.156750: E tensorflow/c/c_api.cc:2184] Already exists: TensorFlow device (GPU:0) is being mapped to multiple CUDA devices (1 now, and 0 previously), which is not supported. This may be the result of providing different GPU configurations (ConfigProto.gpu_options, for example different visible_device_list) when creating multiple Sessions in the same process. This is not  currently supported, see https://github.com/tensorflow/tensorflow/issues/19083
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/client/session.py", line 1596, in __init__
    super(Session, self).__init__(target, graph, config=config)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/client/session.py", line 711, in __init__
    self._session = tf_session.TF_NewSessionRef(self._graph._c_graph, opts)
tensorflow.python.framework.errors_impl.AlreadyExistsError: TensorFlow device (GPU:0) is being mapped to multiple CUDA devices (1 now, and 0 previously), which is not supported. This may be the result of providing different GPU configurations (ConfigProto.gpu_options, for example different visible_device_list) when creating multiple Sessions in the same process. This is not  currently supported, see https://github.com/tensorflow/tensorflow/issues/19083
```

Reply:
You are right. I notice tensorflow/tensorflow#8021